core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -167,25 +125,51 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
 
   @Override
   public void perform() {
-    if (!connected) {
-      return; // Don't waste any time converting images if there's no dashboard connected
-    }
-
-    if (inputSocket.getValue().get().empty()) {
+    final long now = System.nanoTime(); // NOPMD
+    opencv_core.Mat input = inputSocket.getValue().get();
+    if (input.empty() || input.isNull()) {
       throw new IllegalArgumentException("Input image must not be empty");
     }
 
-    synchronized (imageLock) {
-      imencode(".jpeg", inputSocket.getValue().get(), imagePointer,
-          new IntPointer(CV_IMWRITE_JPEG_QUALITY, qualitySocket.getValue().get().intValue()));
-      imageLock.notifyAll();
+    copyJavaCvToOpenCvMat(input, publishMat);
+    serverSource.putFrame(publishMat);
+    if (lastFrame != -1) {
+      long dt = now - lastFrame;
+      serverSource.setFPS((int) (1e9 / dt));
     }
+    lastFrame = now;
+    server.setSource(serverSource);
   }
 
   @Override
   public synchronized void cleanUp() {
     // Stop the video server if there are no Publish Video steps left
-    serverThread.interrupt();
     numSteps--;
+    availablePorts.addFirst(server.getPort());
+    ourTable.getKeys().forEach(ourTable::delete);
   }
+
+  private void copyJavaCvToOpenCvMat(opencv_core.Mat javaCvMat, Mat openCvMat) {
+    if (javaCvMat.depth() != CV_8U && javaCvMat.depth() != CV_8S) {
+      throw new IllegalArgumentException("Only 8-bit depth images are supported");
+    }
+
+    final opencv_core.Size size = javaCvMat.size();
+
+    // Make sure the output resolution is up to date
+    serverSource.setResolution(size.width(), size.height());
+
+    // Make the OpenCV Mat object point to the same block of memory as the JavaCV object.
+    // This requires no data transfers or copies and is O(1) instead of O(n)
+    if (javaCvMat.address() != openCvMat.nativeObj) {
+      try {
+        Field nativeObjField = Mat.class.getField("nativeObj");
+        nativeObjField.setAccessible(true);
+        nativeObjField.setLong(openCvMat, javaCvMat.address());
```

Comment:
This seems like a terrible idea, but I understand what you are doing here.
Do you actually see speed benefits from doing this?
@saudet Is there any reason why doing this assignment could shoot us in the foot?

This is taking the memory address of the OpenCV mat and basically telling JavaCV to use it's memory instead.

Reply:
First it'd say, what is missing from JavaCV that you need to use the other set of bindings? IMO, we should fix that so we don't need to do things like that.

And even when that's necessary, we don't need to use reflection.

Just passing the pointer address around is sufficient, as long as you make sure they are from the same version of OpenCV, compiled with the same compiler, the same flags, etc and to make sure memory isn't prematurely deallocated and that it gets deallocated by the same binaries used to allocate it.

Reply:
Is there a way of creating a `Mat` just using the memory address??

Reply:
@SamCarlberg Ping on this? Thoughts on making sure the memory isn't inconsistently GCed?

Reply:
See my comment at the bottom. In the rewrite, all the memory mapping between different opencv versions will be gone. There will be no wpilib opencv interface anymore.

Reply:
As an example, The following class would be added to GRIP. It will use the JavaCV version of OpenCV, and always copy from the raw image into the requested Mat. It is memory safe, and does not depend at all on the internal representation of how an OpenCV Mat looks. There's some tricks on the cscore jni side to avoid allocations when needed, but everything is in spec, no reflection or undefined behavior.

https://github.com/wpilibsuite/allwpilib/blob/13e15663b6570028a9222b4990d141d16fda3a72/cscore/src/dev/java/edu/wpi/cscore/RawCVMatSink.java

Reply:
@SamCarlberg Do we want to wait for that update then?

Reply:
I think so. Thad's offered to do do the rewrite with the new cscore stuff. Maybe we should also use the cameraserver library for doing the stream hosting instead of manually doing the Networktable side

Reply:
I wouldn't be able to bring it in directly (because of the opencv dependency) but I could bring in most of it. There's a lot extra we wouldnt need there as well.

Reply:
I wouldn't be able to bring it in directly (because of the opencv dependency) but I could bring in most of it. There's a lot extra we wouldnt need there as well.